src/wayland/compositor/transaction.rs

```diff
@@ -0,0 +1,248 @@
+    }
+
+    pub(crate) fn is_same_as(&self, other: &PendingTransaction) -> bool {
+        let ptr1 = self.with_inner_state(|state| state as *const _);
+        let ptr2 = other.with_inner_state(|state| state as *const _);
+        ptr1 == ptr2
+    }
+
+    pub(crate) fn merge_into(&self, into: &PendingTransaction) {
+        if self.is_same_as(into) {
+            // nothing to do
+            return;
+        }
+        // extract our pending surfaces and change our link
+        let mut next = self.inner.clone();
+        let my_state;
+        loop {
+            let tmp = {
+                let mut guard = next.lock().unwrap();
+                match *guard {
+                    TransactionInner::Data(ref mut state) => {
+                        my_state = std::mem::take(state);
+                        *guard = TransactionInner::Fused(into.inner.clone());
+                        break;
+                    }
+                    TransactionInner::Fused(ref into) => into.clone(),
+                }
+            };
+            next = tmp;
+        }
+        // fuse our surfaces into our new transaction state
+        self.with_inner_state(|state| {
+            for (s, id) in my_state.surfaces {
+                if let Some(place) = state.surfaces.iter_mut().find(|place| place.0 == s) {
+                    // the surface is already in the list, update the serial
+                    if place.1 < id {
+                        place.1 = id;
+                    }
+                } else {
+                    // the surface is not in the list, insert it
+                    state.surfaces.push((s, id));
+                }
+            }
+            state.blockers.extend(my_state.blockers);
+        });
+    }
+
+    pub(crate) fn finalize(mut self) -> Transaction {
+        // When finalizing a transaction, this *must* be the last handle to this transaction
+        loop {
+            let inner = match Arc::try_unwrap(self.inner) {
+                Ok(mutex) => mutex.into_inner().unwrap(),
+                Err(_) => panic!("Attempting to finalize a transaction but handle is not the last."),
+            };
+            match inner {
+                TransactionInner::Data(TransactionState {
+                    surfaces, blockers, ..
+                }) => return Transaction { surfaces, blockers },
+                TransactionInner::Fused(into) => self.inner = into,
+            }
+        }
+    }
+}
+pub(crate) struct Transaction {
+    surfaces: Vec<(WlSurface, Serial)>,
+    blockers: Vec<Box<dyn Blocker + Send>>,
+}
+
+impl Transaction {
+    pub(crate) fn state(&self) -> BlockerState {
+        use BlockerState::*;
+        self.blockers
+            .iter()
+            .fold(Released, |acc, blocker| match (acc, blocker.state()) {
+                (Cancelled, _) | (_, Cancelled) => Cancelled,
+                (Pending, _) | (_, Pending) => Pending,
+                (Released, Released) => Released,
+            })
```

Comment:
Couldn't this be a `find` call? Looks cleaner and easier to understand for me.

```suggestion
            .map(|blocker| blocker.state())
            .find(|state| state == Cancelled || state == Pending)
            .unwrap_or(Released)
```

Reply:
That'd make the logic slightly different. Your proposal would early return on the first `Cancelled` or `Pending`, while if the list contains both a `Pending` and a `Cancelled`, my code will always return `Cancelled`.

Reply:
Ah right, but then this needs a comment. I would not have noticed this detail without you pointing it out.